File: app/src/main/java/com/supercilex/robotscouter/data/model/Scout.java
package com.supercilex.robotscouter.data.model;

import android.support.annotation.Keep;
import android.text.TextUtils;

import com.google.firebase.database.Exclude;
import com.supercilex.robotscouter.data.model.metrics.ScoutMetric;

import java.util.ArrayList;
import java.util.List;

public class Scout {
    @Exclude private String mName;

    @Exclude private List<ScoutMetric> mMetrics = new ArrayList<>();

    @Keep
    public String getName() {
        return mName;
    }

    @Keep
    public void setName(String name) {
        mName = name;
    }

    @Exclude
    public List<ScoutMetric> getMetrics() {
        return new ArrayList<>(mMetrics);
    }

    public void add(ScoutMetric metric) {
        mMetrics.add(metric);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Scout scout = (Scout) o;

        return TextUtils.equals(mName, scout.mName) && mMetrics.equals(scout.mMetrics);
    }

    @Override
    public int hashCode() {
        int result = mName == null ? 0 : mName.hashCode();
        result = 31 * result + mMetrics.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return "Scout{" +
                "mName='" + mName + '\'' +
                ", mMetrics=" + mMetrics +
                '}';
    }
}

File: app/src/main/java/com/supercilex/robotscouter/data/model/Scout.kt
package com.supercilex.robotscouter.data.model

import android.support.annotation.Keep
import com.google.firebase.firestore.Exclude
import java.util.Date

data class Scout(
        @Exclude
        @get:Exclude
        val id: String,

        @Exclude
        @get:Keep
        val templateId: String,

        @Exclude
        @get:Keep
        val name: String? = null,

        @Exclude
        @get:Keep
        val timestamp: Date = Date(),

        @Exclude
        @get:Exclude
        val metrics: List<Metric<*>> = emptyList()
)
